### ISSUE TITLE
Unable to use configmap mounted as wp-config.php due to chmod error

### ISSUE THREAD

**Issue by lwj5** [bitnami/bitnami-docker-wordpress#240]
**Description**
I'm trying to generate a config map using a .yaml template defining a wp-config.php key with the value containing the actual configs. And mount this config map key in /bitnami/worpress/wp-config.php.

However, an error is received as the nami is trying to chmod the file which is read-only. The wp-config.php has sufficient permissions as seen below.
`-r--r----- 1 root 1001 4117 May  3 06:55 wp-config.php`

The related issue (https://github.com/bitnami/charts/issues/2473) and PR (https://github.com/bitnami/charts/pull/2491)

**Steps to reproduce the issue:**

1. mount configmap as wp-config.php in /bitnami/wordpress/wp-config.php
2. launch container
3. crash

**Describe the results you received:**
```
wordpre TRACE [prepareDataToPersist] Preparing /opt/bitnami/wordpress/wp-content to persist in /bitnami/wordpress/wp-content
wordpre TRACE [prepareDataToPersist] Preparing /opt/bitnami/wordpress/wp-config.php to persist in /bitnami/wordpress/wp-config.php
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 19 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 18 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 17 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 16 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 15 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 14 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 13 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 12 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 11 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 10 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 9 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 8 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 7 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 6 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 5 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 4 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 3 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 2 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 1 remaining attempts
wordpre TRACE [configurePermissions] File to chmod: /bitnami/wordpress/wp-config.php
wordpre TRACE [retry] Trying to configure permissions... 0 remaining attempts
Error executing 'postInstallation': EROFS: read-only file system, chmod '/bitnami/wordpress/wp-config.php'
nami    TRACE Error: Error executing 'postInstallation': EROFS: read-only file system, chmod '/bitnami/wordpress/wp-config.php'
    at Object.fs.chmodSync (fs.js:1123:18)
    at Object.chmodSync (/opt/bitnami/nami/node_modules/graceful-fs/polyfills.js:248:19)
    at _chmod (/opt/bitnami/nami/node_modules/nami-utils/lib/file/chmod.js:13:6)
    at Object.chmod (/opt/bitnami/nami/node_modules/nami-utils/lib/file/chmod.js:54:5)
    at Object.wrappedFn (/opt/bitnami/nami/node_modules/nami-utils/lib/function-wrapping.js:191:17)
    at _configurePermissions (/.nami/components/com.bitnami.wordpress/lib/component.js:172:17)
    at utilsFunctions.retry (/.nami/components/com.bitnami.wordpress/lib/component.js:208:13)
    at Object.retry (/.nami/components/com.bitnami.wordpress/lib/utils.js:28:14)
    at _.each (/.nami/components/com.bitnami.wordpress/lib/component.js:200:24)
    at r (/opt/bitnami/nami/node_modules/lodash/index.js:5:348)
```

**Describe the results you expected:**
Nami to check if there are sufficient permissions, if yes, proceed without chmod.

**Additional information you deem important (e.g. issue happens only occasionally):**
Nil

**Version**

- Output of `docker version`:

```
Nil
```

- Output of `docker info`:

```
Nil
```

- Output of `docker-compose version` (if applicable):

```
Nil
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
Nil

**Comment by miguelaeh** [comment]
Hi @lwj5 ,
Thanks for the information, I took a look at the PR, and I would like to know if it would be enough adding support for an existing ConfigMap instead of adding a parameter for each WordPress option.
Regarding the proposed solution I will create an internal task to check the implications of changing this and check if is there any other approach to avoid perform changes on the image.
Thank you very much!

**Comment by lwj5** [comment]
@miguelaeh Do you mean the salts? Yes, all 8 of them below can be removed from values.yaml leaving only 2 new fields. You can use wordpressConfigDefine to define all the 8 different salts instead.
```
wordpressAuthKey
wordpressSecureAuthKey
wordpressLoggedInKey
wordpressNonceKey
wordpressAuthSalt
wordpressSecureAuthSalt
wordpressLoggedInSalt
wordpressNonceSalt
```

By doing this, only two additional fields are retained in values.yaml. This would allow you to define almost any configuration into wp-config.php.
```
wordpressConfigDefine
wordpressConfigBlock
```

However, going this route has security implications as salts are secrets and rightly they should be defined in secrets and not in the configmap. Hence my take on putting all the salts into the values.yaml file and passing the secrets as env. But the alternative is doable as well.

To the question of using an existing configmap. As long as any new user uses the wp-config.php outline in the PR with `get_env()`, they will be fine. If not it would be back to the problem of manually change the configmap and also the helm values for any configuration change.

**Comment by miguelaeh** [comment]
>If not, we are right back to the problem of manually changing both the configmap and also helm values for any configuration change (for e.g. db password) as they are not in sync.

I think this is something that can be fixed with https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

**Comment by lwj5** [comment]
@miguelaeh this is not the problem. WP is stateless, so any changes to the configmap will propagate immediately. There is no need for any of the tricks above.

The problem lies in needing to change for example the db password, db host, wordpressScheme so on and so forth. 

One would expect that by doing a 
`helm upgrade chart-name . --set externalDatabase.password=newpass` 
or
`helm upgrade chart-name . --set wordpressScheme=https`

would automatically change the db password or the Wordpress scheme. But it doesn't. You will have to manually change the wp-config.php again.

This way of doing things twice is really redundant. I could have just used the config map with the get_env() variables and the wp-config.php is automatically changed. 

Do you see the point?

**Comment by miguelaeh** [comment]
Hi @lwj5 ,
I see your point. This is something that happens in more cases, but not trivial to solve.
The problem is that we are storing the configurations in the PV using the first installation parameters. When you change those parameters and make an upgrade, the PV is not changed, so the configuration remains the same. This is very common in databases too, where the password is stored into the PV and you can perform the upgrade and see that, the password in the secret changes, but the password in the PV is the same so the database password is the same.
This is something we are aware off and we have it in our to investigate list.
BTW thank you very much for the PR!!

**Comment by morarucostel** [comment]
Hi, any updates on this issue?
I'm trying to setup WP with multi-site configuration, which asks me to update the `wp-config.php` and add some define clauses in a specific place, so I need to be able to use my own wp-config file. 

Is there a word-around to provide your own wp-config file configuration ?

10x

**Comment by miguelaeh** [comment]
Hi,
I am afraid we didn't have the time to work on this. We will update the threads once we have news.
Sorry for the inconvenience.

**Comment by lwj5** [comment]
@miguelaeh is the wordpress installer package open source? where can i take a look at it

**Comment by miguelaeh** [comment]
Hi @lwj5 ,
It is publicly hosted in any repo but you can see the code under ` ~/.nami/components/com.bitnami.wordpress/` inside the docker image.
We are in the process of migrating all our assets from nodejs to bash to make the code directly available on GitHub

**Comment by marcosbc** [comment]
Hi @lwj5, we have just released tag [`5.7.1-debian-10-r21`](https://github.com/bitnami/bitnami-docker-wordpress#571-debian-10-r21) of the container images which implements a lot of new features. In particular it adds configuration specified via `WORDPRESS_EXTRA_WP_CONFIG_CONTENT` right before the "That's all, stop editing!" comment.

> The configuration logic is now based on Bash scripts in the *rootfs/* folder.

You can check the configuration scripts [here](https://github.com/bitnami/bitnami-docker-wordpress/tree/master/5/debian-10/rootfs/opt/bitnami/scripts).

If that is not enough, you should be able to mount the `wp-config.php` with the new image file by specifying the following environment variable (so that `wp-config.php` is not persisted):

```yaml
- WORDPRESS_DATA_TO_PERSIST=wp-content
```
